python/cog/server/worker.py

```diff
@@ -45,48 +49,75 @@ def __init__(self, predictor_ref: str, tee_output: bool = True) -> None:
         self._state = WorkerState.NEW
         self._allow_cancel = False
 
+        self._result: Optional["Future[Done]"] = None
+        self._subscribers: Dict[int, Optional[Callable[[_PublicEventType], None]]] = {}
+
+        self._pool = ThreadPoolExecutor(max_workers=1)
```

Comment:
There is exactly as much threading as in `main`, namely one worker thread, but in a more sensible place. In the new code the single extra thread can in principle handle events for multiple predictions at once, which is not true of PredictionRunner's use of threads. That's the entire point of this PR.

The only alternative to multithreading here is async, and I think I've been pretty clear that I don't think trying to convert the entirety of Cog to `async` all at once is a good idea. It may well be the best thing to do long term, but the "viral" nature of async signatures is such that the _only_ way to avoid multithreading is if the entire application supports cooperative multitasking. That involves changing _much_ more code than this PR, which is why we've taken this approach.